xds/src/main/java/io/grpc/xds/XdsNameResolver.java

```diff
@@ -635,89 +633,71 @@ public Result selectConfig(PickSubchannelArgs args) {
+    private final String authority;
+    private final XdsDependencyManager xdsDependencyManager;
     private boolean stopped;
     @Nullable
     private Set<String> existingClusters;  // clusters to which new requests can be routed
-    @Nullable
-    private RouteDiscoveryState routeDiscoveryState;
+    private XdsConfig lastConfig;
 
-    ResolveState(String ldsResourceName) {
-      this.ldsResourceName = ldsResourceName;
+    private ResolveState(String ldsResourceName) {
+      authority = overrideAuthority != null ? overrideAuthority : encodedServiceAuthority;
+      xdsDependencyManager =
+          new XdsDependencyManager(xdsClient, this, syncContext, authority, ldsResourceName,
+              nameResolverArgs, scheduler);
     }
 
-    @Override
-    public void onChanged(final XdsListenerResource.LdsUpdate update) {
+    private void shutdown() {
       if (stopped) {
         return;
       }
-      logger.log(XdsLogLevel.INFO, "Receive LDS resource update: {0}", update);
-      HttpConnectionManager httpConnectionManager = update.httpConnectionManager();
-      List<VirtualHost> virtualHosts = httpConnectionManager.virtualHosts();
-      String rdsName = httpConnectionManager.rdsName();
-      cleanUpRouteDiscoveryState();
-      if (virtualHosts != null) {
-        updateRoutes(virtualHosts, httpConnectionManager.httpMaxStreamDurationNano(),
-            httpConnectionManager.httpFilterConfigs());
-      } else {
-        routeDiscoveryState = new RouteDiscoveryState(
-            rdsName, httpConnectionManager.httpMaxStreamDurationNano(),
-            httpConnectionManager.httpFilterConfigs());
-        logger.log(XdsLogLevel.INFO, "Start watching RDS resource {0}", rdsName);
-        xdsClient.watchXdsResource(XdsRouteConfigureResource.getInstance(),
-            rdsName, routeDiscoveryState, syncContext);
-      }
+
+      stopped = true;
+      xdsDependencyManager.shutdown();
     }
 
     @Override
-    public void onError(final Status error) {
-      if (stopped || receivedConfig) {
+    public void onUpdate(XdsConfig update) {
+      if (stopped) {
         return;
       }
-      listener.onError(Status.UNAVAILABLE.withCause(error.getCause()).withDescription(
-          String.format("Unable to load LDS %s. xDS server returned: %s: %s",
-          ldsResourceName, error.getCode(), error.getDescription())));
+      logger.log(XdsLogLevel.INFO, "Receive XDS resource update: {0}", update);
+      lastConfig = update;
+
+      // Process Route
+      HttpConnectionManager httpConnectionManager = update.getListener().httpConnectionManager();
+      VirtualHost virtualHost = update.getVirtualHost();
+
+      updateRoutes(virtualHost, httpConnectionManager.httpMaxStreamDurationNano(),
+          httpConnectionManager.httpFilterConfigs());
     }
 
     @Override
-    public void onResourceDoesNotExist(final String resourceName) {
+    public void onError(String resourceContext, Status error) {
```

Comment:
As of https://github.com/grpc/proposal/pull/482 doesn't onError and onDoesNotExist go away?

Reply:
Yep. I had just sent out this PR before that gRFC update. The behavior here already matches, but I do need to update the APIs.